src/main/java/at/ac/tuwien/kr/alpha/grounder/Substitution.java

```diff
@@ -203,13 +289,22 @@ private static Substitution specializeSubstitution(Substitution substitution, Te
 			Term substitutedGeneralTerm = substitution.eval((VariableTerm) generalTerm);
 			// If the variable is not bound already, bind it to the specific term.
 			if (substitutedGeneralTerm == null) {
-				substitution.put((VariableTerm) generalTerm, specificTerm);
+				substitution.put((VariableTerm) generalTerm, specificTerm.substitute(substitution));
+				// TODO: check in case we have X -> Y already and are adding Y -> c, if X -> Y must be replaced with X -> c.
+				for (Map.Entry<VariableTerm, Term> variableSubstitution : substitution.substitution.entrySet()) {
+					if (variableSubstitution.getValue() instanceof VariableTerm
+						&& variableSubstitution.getValue() == generalTerm) {
+						System.out.println("Problem.");
+					}
+				}
 				return substitution;
 			}
 			// The variable is bound, check whether its result is exactly the specific term.
 			// Note: checking whether the bounded term is more general than the specific one would yield
 			//       wrong results, e.g.: p(X,X) and p(f(A),f(g(B))) are incomparable, but f(A) is more general than f(g(B)).
-			if (substitutedGeneralTerm != specificTerm) {
+			if (substitutedGeneralTerm == specificTerm) {
```

Comment:
Since our terms all use an `Interner` and should only be constructable by `getInstance(..)` methods, it is safe to assume that two terms are equal if-and-only-if they are the same object.